torch/csrc/jit/codegen/cuda/arith.cpp

```diff
@@ -553,6 +555,47 @@ TensorView* index_select(TensorView* lookup_tv, int dim, TensorView* index_tv) {
   return out;
 }
 
+// torch.gather
+TensorView* torch_gather(TensorView* inp, int dim, TensorView* index) {
+  auto inp_domain = TensorDomain::noReductions(inp->getMaybeRFactorDomain());
+  auto idx_domain = TensorDomain::noReductions(index->getMaybeRFactorDomain());
+  TORCH_CHECK(
+      inp_domain.size() > 0, "torch.gather can not be applied to 0d tensor.");
+  TORCH_CHECK(
+      idx_domain.size() == inp_domain.size(),
+      "the input and index tensor must have the same dimensions for torch.gather");
+
+  if (dim < 0) {
+    dim += idx_domain.size();
+  }
+  TORCH_CHECK(
+      dim >= 0 && dim < inp_domain.size(),
+      "torch.gather on invalid axis, received: ",
+      dim,
+      " however tensor view only has ",
+      inp_domain.size(),
+      " non-reduction dims.");
+  std::vector<IterDomain*> out_domain;
+  for (int i = 0; i < idx_domain.size(); ++i) {
+    out_domain.push_back(
+        IterDomainBuilder(idx_domain[i])
+            .iter_type(
+                idx_domain[i]->getIterType() == IterType::Iteration
+                    ? IterType::GatherScatter
+                    : idx_domain[i]->getIterType())
```

Comment:
I suppose this is needed for a case where the index tensor has broadcast domains.

Can the selected axis of the index tensor be a broadcast? I think it can. Pinging @zasdfgbnm

Reply:
When one of the dimension sizes of index tensor is equal one, it will become broadcast domain.

Reply:
Yeah, but then it's a little inconsistent that sometimes we have `GatherScatter` IterDomains but sometimes `Broadcast`. I still believe we really don't need `GatherScatter`, but removing it would be a different cleanup PR.

Reply:
Yeah, I will explore methods that don't require `GatherScatter`.